### PR TITLE
provider/aws: Add support for api_gateway_account

### PR DESCRIPTION
**Pull request by radeksimko**
<del>I've reached out to Amazon to get a confirmation whether it's really impossible to reset CloudWatch Role ARN to empty string. That's why I marked this as WIP, otherwise it works ok (within the limitations of the API).</del> see below

### Test plan

```sh
make testacc TEST=./builtin/providers/aws TESTARGS='-run=AWSAPIGatewayAccount_basic'
```
```
TF_ACC=1 go test ./builtin/providers/aws -v -run=AWSAPIGatewayAccount_basic -timeout 120m
=== RUN   TestAccAWSAPIGatewayAccount_basic
--- PASS: TestAccAWSAPIGatewayAccount_basic (61.86s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	61.880s
```

### CODE REVIEW

**Comment by radeksimko** [comment]
I received response from AWS support confirming my findings - i.e. it's not possible to remove an ARN or make it blank. 😢 

> Hi,

> From my testing it doesn't look like you can remove this. While this shouldn't be a huge security issue as it's only logging events within that particular APIGW region, I will forward this to our APIGW team to look at as a feature request moving forward.

This is therefore ready for review/merge.

File: builtin/providers/aws/resource_aws_api_gateway_account.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		},
+	}
+}
+
+func resourceAwsApiGatewayAccountRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	log.Printf("[INFO] Reading API Gateway Account %s", d.Id())
+	account, err := conn.GetAccount(&apigateway.GetAccountInput{})
+	if err != nil {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: Is there a chance that someone deleting an ApiGatewayAccount (using the AWS Console) could cause an error here? Should we check for a 404 and remove from state if so?

File: builtin/providers/aws/resource_aws_api_gateway_account.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			// even though that's default settings for new AWS accounts
+			// BadRequestException: The role ARN is not well formed
+			operations = append(operations, &apigateway.PatchOperation{
+				Op:    aws.String("replace"),
+				Path:  aws.String("/cloudwatchRoleArn"),
+				Value: aws.String(arn),
+			})
+		}
+	}
+	input.PatchOperations = operations
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: If operations is empty, what will happen to the update request?

File: builtin/providers/aws/resource_aws_api_gateway_account.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Updating API Gateway Account failed: %s", err)
+	}
+	log.Printf("[DEBUG] API Gateway Account updated: %s", out)
+
+	d.SetId("api-gateway-account")
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: Why does the SetId only happen as part of Update? Does this mean we have no Id before the update?

File: builtin/providers/aws/resource_aws_api_gateway_account.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	return resourceAwsApiGatewayAccountRead(d, meta)
+}
+
+func resourceAwsApiGatewayAccountDelete(d *schema.ResourceData, meta interface{}) error {
+	// There is no API for "deleting" account or resetting it to "default" settings
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: Will this leak resources?

File: builtin/providers/aws/structure.go [hunk 5]
```diff
@@ -1097,3 +1097,22 @@ func sortInterfaceSlice(in []interface{}) []interface{} {
 
 	return b
 }
+
+func flattenApiGatewayThrottleSettings(settings *apigateway.ThrottleSettings) []map[string]interface{} {
```
Comment (stack72) [inline builtin/providers/aws/structure.go]: Would love to see a test for this

**Comment by stack72** [comment]
Few small questions about the PR and 1 nit pick (about adding a test) but that can be fixed later if needed :)

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 1 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: There is no way to delete the "account" even through the AWS Console... so no.

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 3 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: Technically I only `SetId` to "create" the resource inside tfstate... there cannot be multiple API Gateway "accounts" or anything like that.

Otherwise there can be existing settings, which would be region-wide and this API call overrides that.

Also note that this method is used for both Create & Update.

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 4 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: Well technically yes, but there is no way to prevent this on the API level nor in the AWS Console. I can double check but I think that CloudFormation must be doing exactly the same thing with http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigateway-account.html

File: builtin/providers/aws/structure.go (hunk 5 above)
Comment (radeksimko) [inline builtin/providers/aws/structure.go]: Good point, I can definitely add one.

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 2 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: It will send an empty update API call, which seems to be valid (or at least no error is being returned in such case). I can add a test for it.

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 4 above)
Comment (stack72) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: This is something we can figure out later. I am just not sure how the resources will affect us - like if there are limits, costs etc.

**Comment by stack72** [comment]
I guess we can merge this for now and then figure out the leaked resources part later? The tests pass (As below) so this LGTM!

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=AWSAPIGatewayAccount_basic'
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /vendor/)
TF_ACC=1 go test ./builtin/providers/aws -v -run=AWSAPIGatewayAccount_basic -timeout 120m
=== RUN   TestAccAWSAPIGatewayAccount_basic
--- PASS: TestAccAWSAPIGatewayAccount_basic (57.28s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	57.298s
```

File: builtin/providers/aws/resource_aws_api_gateway_account.go (hunk 4 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_account.go]: I believe there are no costs belonging specifically to this resource, but I share your concerns - that's why I raised this with AWS. Until they implement "remove/delete" API method, there isn't much we can do on our side though.

**Comment by radeksimko** [comment]
Let me just add two tests (empty resource with "empty" update API call + `flattenApiGatewayThrottleSettings`) and then you can give it final review & merge.

Hopefully I'll get to it tomorrow.

**Comment by stack72** [comment]
👍

**Comment by radeksimko** [comment]
Both tests added per comments. This is now ready for final review @stack72

**Comment by stack72** [comment]
@radeksimko this is awesome :) Thanks!
